File: code/88.merge-sorted-array.py
#
# @lc app=leetcode id=88 lang=python3
#
# [88] Merge Sorted Array
#

# @lc code=start
class Solution:
    def merge(self, nums1: List[int], m: int, nums2: List[int], n: int) -> None:
        """
        Do not return anything, modify nums1 in-place instead.
        """
        if m == 0:
            for i in range(n):
                nums1[i] = nums2[i]
            return
        elif n == 0:
            return

        mark = [0] * (m+1)
        i = 0
        j = 0
        while i < m and j < n:
            if nums2[j] < nums1[i]:
                mark[i] = mark[i] + 1
                j = j + 1
            else:
                i = i + 1
        while j < n:
            mark[m] = mark[m] + 1
            j = j + 1
        
        i = m - 1
        j = n - 1
        k = m + n - 1
        while k >= 0 and i >= 0:
            for _ in range(mark[i+1]):
                nums1[k] = nums2[j]
                j = j - 1
                k = k - 1
            nums1[k] = nums1[i]
            i = i - 1
            k = k - 1
        for _ in range(mark[0]):
            nums1[k] = nums2[j]
            j = j - 1
            k = k - 1
# @lc code=end

File: code/88.merge-sorted-array.cpp
/*
 * @lc app=leetcode id=88 lang=cpp
 *
 * [88] Merge Sorted Array
 */

class Solution {
public:
    void merge(vector<int>& nums1, int m, vector<int>& nums2, int n) {
        int index1 = m - 1;
        int index2 = n - 1;
        int index3 = m + n - 1;

        while(index1 >= 0 && index2 >= 0) {
            if(nums1[index1] > nums2[index2]) {
                nums1[index3] = nums1[index1];
                --index1;
            }
            else {
                nums1[index3] = nums2[index2];
                --index2;
            }
            --index3;
        }
        while(index2 >= 0) {
            nums1[index3] = nums2[index2];
            --index2;
            --index3;
        }
    }
};

